cnd/src/swap_protocols/ledger/bitcoin.rs

```diff
@@ -1,56 +1,50 @@
 use crate::comit_api::LedgerKind;
 
-#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
-pub struct Mainnet;
-
-#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
-pub struct Testnet;
-
-#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
-pub struct Regtest;
+macro_rules! declare_ledger {
+    ($ledger:ident) => {
+        #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+        pub struct $ledger;
+    };
+}
+declare_ledger!(Mainnet);
+declare_ledger!(Testnet);
+declare_ledger!(Regtest);
 
 pub trait Bitcoin:
     Sized + std::fmt::Debug + std::hash::Hash + Eq + Sync + Copy + Send + Into<LedgerKind> + 'static
 {
 }
 
-pub trait Network {
-    fn network() -> ::bitcoin::Network;
-}
-
 impl Bitcoin for Mainnet {}
-
-impl From<Mainnet> for LedgerKind {
-    fn from(_: Mainnet) -> Self {
-        LedgerKind::BitcoinMainnet
-    }
-}
-impl Network for Mainnet {
-    fn network() -> ::bitcoin::Network {
-        ::bitcoin::Network::Bitcoin
-    }
-}
 impl Bitcoin for Testnet {}
-
-impl From<Testnet> for LedgerKind {
-    fn from(_: Testnet) -> Self {
-        LedgerKind::BitcoinTestnet
-    }
-}
-impl Network for Testnet {
-    fn network() -> ::bitcoin::Network {
-        ::bitcoin::Network::Testnet
-    }
-}
-
 impl Bitcoin for Regtest {}
-impl From<Regtest> for LedgerKind {
-    fn from(_: Regtest) -> Self {
-        LedgerKind::BitcoinRegtest
-    }
-}
-impl Network for Regtest {
-    fn network() -> ::bitcoin::Network {
-        ::bitcoin::Network::Regtest
-    }
-}
+
+pub trait Network {
+    fn network() -> bitcoin::Network;
+}
+
+macro_rules! impl_network {
+    ($ledger:ty, $network:expr) => {
+        impl Network for $ledger {
+            fn network() -> bitcoin::Network {
+                $network
+            }
+        }
+    };
+}
+impl_network!(Mainnet, bitcoin::Network::Bitcoin);
+impl_network!(Testnet, bitcoin::Network::Testnet);
+impl_network!(Regtest, bitcoin::Network::Regtest);
+
+macro_rules! impl_from_for_ledgerkind {
+    ($ledger:ty, $ledger_kind:expr) => {
+        impl From<$ledger> for LedgerKind {
+            fn from(_: Mainnet) -> Self {
+                $ledger_kind
+            }
+        }
+    };
+}
+impl_from_for_ledgerkind!(Mainnet, LedgerKind::BitcoinMainnet);
```

Comment:
This impl is at the wrong place it should be in `comit_api` as it is where `LedgerKind` is defined.
`swap_protocols` is part of the comit lib, imported by cnd.

Reply:
lols, you put these here @D4nte :)

Reply:
I know :)